File: frontend/src/components/benchmark/content/k_core_decomposition/implementations/k_core_decomposition.go
package main

type KCoreDecomposition struct {
    adjacency map[int]map[int]bool
    nodes     map[int]bool
}

func NewKCoreDecomposition() *KCoreDecomposition {
    return &KCoreDecomposition{
        adjacency: make(map[int]map[int]bool),
        nodes:     make(map[int]bool),
    }
}

func (kc *KCoreDecomposition) AddEdge(u, v int) {
    if kc.adjacency[u] == nil {
        kc.adjacency[u] = make(map[int]bool)
    }
    if kc.adjacency[v] == nil {
        kc.adjacency[v] = make(map[int]bool)
    }
    kc.adjacency[u][v] = true
    kc.adjacency[v][u] = true
    kc.nodes[u] = true
    kc.nodes[v] = true
}

func (kc *KCoreDecomposition) BuildFromEdges(edges [][2]int) {
    for _, edge := range edges {
        kc.AddEdge(edge[0], edge[1])
    }
}

func (kc *KCoreDecomposition) KCoreDecomposition() map[int]int {
    core := make(map[int]int)
    degree := make(map[int]int)

    for node := range kc.nodes {
        degree[node] = len(kc.adjacency[node])
    }

    buckets := make([][]int, len(kc.nodes)+1)
    for i := range buckets {
        buckets[i] = []int{}
    }

    maxDegree := 0
    for node := range kc.nodes {
        buckets[degree[node]] = append(buckets[degree[node]], node)
        if degree[node] > maxDegree {
            maxDegree = degree[node]
        }
    }

    removed := make(map[int]bool)
    k := 0

    for i := 0; i <= maxDegree; i++ {
        for _, node := range buckets[i] {
            if removed[node] {
                continue
            }

            if i > k {
                k = i
            }
            core[node] = k
            removed[node] = true

            for neighbor := range kc.adjacency[node] {
                if !removed[neighbor] {
                    degree[neighbor]--
                    if degree[neighbor] <= i {
                        buckets[degree[neighbor]] = append(buckets[degree[neighbor]], neighbor)
                    }
                }
            }
        }
    }

    for node := range kc.nodes {
        if _, exists := core[node]; !exists {
            core[node] = 0
        }
    }

    return core
}

func (kc *KCoreDecomposition) GetKCore(k int) []int {
    core := kc.KCoreDecomposition()
    result := []int{}

    for node := range kc.nodes {
        if core[node] >= k {
            result = append(result, node)
        }
    }

    return result
}

File: frontend/src/components/benchmark/content/k_core_decomposition/implementations/k_core_decomposition.php
<?php

class KCoreDecomposition {
    private array $adjacency = [];
    private array $nodes = [];

    public function addEdge(int $u, int $v): void {
        if (!isset($this->adjacency[$u])) {
            $this->adjacency[$u] = [];
        }
        if (!isset($this->adjacency[$v])) {
            $this->adjacency[$v] = [];
        }
        $this->adjacency[$u][$v] = true;
        $this->adjacency[$v][$u] = true;
        $this->nodes[$u] = true;
        $this->nodes[$v] = true;
    }

    public function buildFromEdges(array $edges): void {
        foreach ($edges as $edge) {
            $this->addEdge($edge[0], $edge[1]);
        }
    }

    public function kCoreDecomposition(): array {
        $core = [];
        $degree = [];

        foreach (array_keys($this->nodes) as $node) {
            $degree[$node] = count($this->adjacency[$node]);
        }

        $buckets = array_fill(0, count($this->nodes) + 1, []);
        $maxDegree = 0;

        foreach (array_keys($this->nodes) as $node) {
            $buckets[$degree[$node]][] = $node;
            $maxDegree = max($maxDegree, $degree[$node]);
        }

        $removed = [];
        $k = 0;

        for ($i = 0; $i <= $maxDegree; $i++) {
            foreach ($buckets[$i] as $node) {
                if (isset($removed[$node])) continue;

                $k = max($k, $i);
                $core[$node] = $k;
                $removed[$node] = true;

                foreach (array_keys($this->adjacency[$node]) as $neighbor) {
                    if (!isset($removed[$neighbor])) {
                        $degree[$neighbor]--;
                        if ($degree[$neighbor] <= $i) {
                            $buckets[$degree[$neighbor]][] = $neighbor;
                        }
                    }
                }
            }
        }

        foreach (array_keys($this->nodes) as $node) {
            if (!isset($core[$node])) {
                $core[$node] = 0;
            }
        }

        return $core;
    }

    public function getKCore(int $k): array {
        $core = $this->kCoreDecomposition();
        $result = [];

        foreach (array_keys($this->nodes) as $node) {
            if ($core[$node] >= $k) {
                $result[] = $node;
            }
        }

        return $result;
    }
}

?>